controllers/vmi_controller.go

```diff
@@ -0,0 +1,212 @@
+package controllers
+
+import (
+	gocontext "context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	kubedrain "k8s.io/kubectl/pkg/drain"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-kubevirt/api/v1alpha1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type VmiReconciler struct {
+	client.Client
+}
+
+func (r *VmiReconciler) SetupWithManager(ctx gocontext.Context, mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&kubevirtv1.VirtualMachineInstance{}).
+		WithEventFilter(predicates.ResourceHasFilterLabel(ctrl.LoggerFrom(ctx), infrav1.KubevirtMachineNameLabel)).
+		Build(r)
+
+	return err
+}
+
+func (r VmiReconciler) Reconcile(ctx gocontext.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	vmi := &kubevirtv1.VirtualMachineInstance{}
+	err := r.Get(ctx, req.NamespacedName, vmi)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(4).Info(fmt.Sprintf("Can't find virtualMachineInstance %s; it was already deleted.", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, fmt.Sprintf("failed to read VMI %s", req.Name))
+		return ctrl.Result{}, err
+	}
+
+	// KubeVirt will set the EvacuationNodeName field in case of guest node eviction. If the field is not set, there is
+	// nothing to do.
+	nodeName := vmi.Status.EvacuationNodeName
+	if len(nodeName) == 0 { // no need to drain
+		logger.V(4).Info(fmt.Sprintf("The virtualMachineInstance %s is not marked for deletion. Nothing to do here", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	cluster, err := r.getCluster(ctx, vmi)
+	if err != nil {
+		logger.Error(err, "Can't get the cluster form the VirtualMachineInstance")
+		return ctrl.Result{}, err
+	}
+
+	nodeDrained, retryDuration, err := r.drainNode(ctx, cluster, nodeName, logger)
```

Comment:
Why there should be cases returning `nodeDrained=false` instead of just returning the error?

Reply:
> Why there should be cases returning `nodeDrained=false` instead of just returning the error?

Because for most cases (but not all of them), we don't want to retry. returning error with the result will cause retry. The calling function can't know which error to return with the result and which error to ignore.